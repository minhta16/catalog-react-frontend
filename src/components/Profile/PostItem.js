import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { Typography, Grid, IconButton, Menu, MenuItem, Link as MUILink } from '@material-ui/core';
import MoreVertIcon from '@material-ui/icons/MoreVert';
import UpdateIcon from '@material-ui/icons/Update';
import DeleteIcon from '@material-ui/icons/Delete';
import { selectCategory } from 'selectors/categories';
import { selectCurrentUserProp } from 'selectors/users';
import { openSnackbar as openSnackbarRedux } from 'actions/misc';
import { deletePostAndRefetch as deletePostAndRefetchRedux } from 'actions/users';
import ConfirmDialog from 'components/Shared/ConfirmDialog';

export class PostItem extends Component {
  state = {
    anchorEl: undefined,
    openConfirm: false,
  };

  handleClick = (e) => {
    this.setState({
      anchorEl: e.target,
    });
  };

  handleClose = () => {
    this.setState({
      anchorEl: undefined,
    });
  };

  /**
   * dispatch a delete action when Confirm is clicked in the delete confirm dialog */

  handleConfirmDelete = async () => {
    const { deletePostAndRefetch, token, post, openSnackbar } = this.props;
    deletePostAndRefetch(token, post.category_id, post.id);
    openSnackbar('Item deleted!');
  };

  /**
   * open the delete dialog when delete button is hit
   */
  toggleDeleteDialog = () => {
    const { openConfirm } = this.state;
    this.setState({
      openConfirm: !openConfirm,
    });
  };

  render() {
    const { post, subtitleLength, category } = this.props;
    const { anchorEl, openConfirm } = this.state;

    // A menu which contains an update and delete button
    const menu = (
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={this.handleClose}
        getContentAnchorEl={null}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      >
        {/* Redirect to edit page if user clicks edit */}
        <MenuItem
          onClick={this.handleClose}
          component={Link}
          exact="true"
          to={`/profile/${post.category_id}/${post.id}/edit`}
        >
          <Grid container alignItems="center">
            <UpdateIcon color="primary" className="icon-padding" />
            <Typography variant="button" color="primary">
              Update
            </Typography>
          </Grid>
        </MenuItem>
        <MenuItem id="post-item-delete" onClick={this.toggleDeleteDialog}>
          <DeleteIcon className="icon-padding" color="error" />
          <Typography variant="button" color="error">
            Delete
          </Typography>
        </MenuItem>
      </Menu>
    );

    return (
      <Grid container justify="space-between" spacing={2}>
        <ConfirmDialog
          open={openConfirm}
          title="Delete"
          onConfirm={this.handleConfirmDelete}
          onCancel={this.toggleDeleteDialog}
          onClose={this.toggleDeleteDialog}
          contentText={`Are you sure you want to delete ${post.name}?`}
        />
        <Grid item>
          <Grid container direction="column">
            {/* Redirect to the post page when user clicked on the name */}
            <MUILink
              variant="h4"
              color="inherit"
              component={Link}
              underline="none"
              exact="true"
              to={`/${post.category_id}/${post.id}`}
            >
              {post.name}
            </MUILink>
            <Typography variant="subtitle1">{`Category: ${category.name}`}</Typography>
            <Typography variant="subtitle1">{`Created: ${post.created}`}</Typography>
            <Typography variant="caption">
              {/* Add triple dots to the end of a description if description is longer than 100 characters */}
              {`Description: ${post.description.substr(0, subtitleLength)}`}
              {post.description.length > subtitleLength && `...`}
            </Typography>
          </Grid>
        </Grid>
        <Grid item>
          <Typography />
          <IconButton onClick={this.handleClick}>
            <MoreVertIcon />
          </IconButton>
        </Grid>
        {menu}
      </Grid>
    );
  }
}

PostItem.propTypes = {
  token: PropTypes.string.isRequired,
  post: PropTypes.object.isRequired,
  subtitleLength: PropTypes.number,
  category: PropTypes.object.isRequired,
  deletePostAndRefetch: PropTypes.func.isRequired,
  openSnackbar: PropTypes.func.isRequired,
};

PostItem.defaultProps = {
  subtitleLength: 100,
};

const mapStateToProps = (state, ownProps) => ({
  category: selectCategory(state, ownProps.post.category_id),
  token: selectCurrentUserProp(state, 'token'),
});

const mapDispatchToProps = {
  deletePostAndRefetch: deletePostAndRefetchRedux,
  openSnackbar: openSnackbarRedux,
};
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(PostItem);
