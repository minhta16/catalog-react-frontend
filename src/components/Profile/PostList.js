import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Typography } from '@material-ui/core';
import { openSnackbar as openSnackbarRedux } from 'actions/misc';
import PostItem from './PostItem';

export class PostList extends Component {
  state = {};

  handleOpenSnackbar = () => {
    const { openSnackbar } = this.props;
    openSnackbar('Post created!');
  };

  render() {
    const { posts } = this.props;
    return (
      <div style={{ margin: '.5rem' }}>
        <Typography variant="h4" className="center-margin">
          Your Posts
        </Typography>
        {posts.length > 0 ? (
          posts.map((post) => <PostItem key={post.id} post={post} />)
        ) : (
          <Typography className="center-margin" variant="body1">
            You have no post. Go ahead and create some. Our community would appreciate your posts!
          </Typography>
        )}
      </div>
    );
  }
}

PostList.propTypes = {
  posts: PropTypes.array.isRequired,
  openSnackbar: PropTypes.func.isRequired,
};

export default connect(
  null,
  { openSnackbar: openSnackbarRedux },
)(PostList);
