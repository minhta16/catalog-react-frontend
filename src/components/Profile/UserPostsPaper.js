import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Typography, Divider, Paper, Avatar, Grid, IconButton } from '@material-ui/core';
import LibraryAddIcon from '@material-ui/icons/LibraryAdd';
import PostList from './PostList';

export function UserPostsPaper(props) {
  const { username, posts } = props;
  return (
    <Paper className="left-margin">
      <Grid container alignItems="center" justify="space-between">
        <Grid item>
          <Grid container alignItems="center">
            <Avatar
              style={{
                backgroundColor: '#fdb600',
                color: '#fff',
                margin: '0.5rem',
              }}
            >
              {username.charAt(0)}
            </Avatar>
            <Typography variant="body1">{username}</Typography>
          </Grid>
        </Grid>
        <Grid item>
          <IconButton color="primary" component={Link} exact="true" to="/new-item">
            <LibraryAddIcon className="icon-padding" />
            <Typography variant="button">New item</Typography>
          </IconButton>
        </Grid>
      </Grid>
      <Divider />
      <PostList posts={posts} />
    </Paper>
  );
}

UserPostsPaper.propTypes = {
  username: PropTypes.string.isRequired,
  posts: PropTypes.array,
};

UserPostsPaper.defaultProps = {
  posts: [],
};

export default UserPostsPaper;
