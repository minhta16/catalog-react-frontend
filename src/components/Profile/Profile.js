import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import UserPostsPaper from 'components/Profile/UserPostsPaper';
import { Container } from '@material-ui/core';
import { fetchCurrentUserPost as fetchCurrentUserPostRedux } from 'actions/users';
import {
  selectCurrentUser,
  selectCurrentUserPosts,
  selectCurrentUserLoading,
} from 'selectors/users';

export class Profile extends Component {
  // Always refetch post if Profile is mounted
  componentDidMount() {
    const { fetchCurrentUserPost, currentUser } = this.props;
    fetchCurrentUserPost(currentUser.token);
  }

  render() {
    const { currentUser, currentUserPosts, currentUserLoading } = this.props;
    return (
      <Container maxWidth="lg">
        <UserPostsPaper
          username={currentUser.username}
          token={currentUser.token}
          posts={currentUserPosts}
          loading={currentUserLoading}
        />
      </Container>
    );
  }
}
Profile.propTypes = {
  currentUser: PropTypes.object.isRequired,
  fetchCurrentUserPost: PropTypes.func.isRequired,
  currentUserPosts: PropTypes.array.isRequired,
  currentUserLoading: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
  currentUser: selectCurrentUser(state),
  currentUserPosts: selectCurrentUserPosts(state, 'reverse'),
  currentUserLoading: selectCurrentUserLoading(state),
});

const mapDispatchToProps = {
  fetchCurrentUserPost: fetchCurrentUserPostRedux,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Profile);
