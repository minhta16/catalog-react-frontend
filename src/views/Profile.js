import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import UserPostsPaper from 'components/UserPostsPaper';
import { Container } from '@material-ui/core';
import { selectCurrentUser } from 'references/redux/selectors/users';

export function Post(props) {
  const { currentUser } = props;
  return (
    <Container maxWidth="lg">
      <UserPostsPaper username={currentUser.username} token={currentUser.token} />
    </Container>
  );
}

Post.propTypes = {
  currentUser: PropTypes.shape.isRequired,
};

const mapStateToProps = (state) => ({
  currentUser: selectCurrentUser(state),
});
Post.propTypes = {};

export default connect(mapStateToProps)(Post);
