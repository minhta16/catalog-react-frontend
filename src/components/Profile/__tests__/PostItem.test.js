/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import { PostItem } from 'components/Profile/PostItem';
import { IconButton, Menu } from '@material-ui/core';

describe('components/PostItem', () => {
  let wrapper;
  let props;
  // let warning;

  // const update = () => {
  //   wrapper.update();
  //   warning = wrapper.find(Warning);
  // };

  const setup = () => {
    wrapper = shallow(<PostItem {...props} />);
  };

  beforeEach(() => {
    props = {
      post: {
        name: 'bro',
        description: 'bro',
        category_id: '1',
        created: '2019',
      },
      token: 'minh',
      category: {
        name: 'cat',
        id: '1',
      },
      deletePostAndRefetch: jest.fn(),
      openSnackbar: jest.fn(),
      subtitleLength: 100,
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render three dots if description is not lengthy', () => {
    props.post.description = '10 characters';
    setup();
    expect(
      wrapper
        .find('[variant="caption"]')
        .text()
        .includes('...'),
    ).toBe(false);
  });

  it('should render three dots if description is to lengthy', () => {
    props.post.description =
      '100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters 100 characters ';
    setup();
    expect(
      wrapper
        .find('[variant="caption"]')
        .text()
        .includes('...'),
    ).toBe(true);
  });

  describe('handleClick', () => {
    it('should set the anchorEl when button is clicked', () => {
      setup();
      const button = wrapper.find(IconButton);
      button.simulate('click', {
        target: button,
      });
      expect(wrapper.state().anchorEl).toBe(button);
    });
  });

  describe('handleClose', () => {
    it('should set the anchorEl when button is clicked', () => {
      setup();
      const menu = wrapper.find(Menu);
      menu.simulate('close');
      expect(wrapper.state().anchorEl).toBe(undefined);
    });
  });

  describe('toggleDeleteDialog', () => {
    it('should set the anchorEl when button is clicked', () => {
      setup();
      const menu = wrapper.find('#post-item-delete');
      const { openConfirm } = wrapper.state();
      menu.simulate('click');
      expect(wrapper.state().openConfirm).toBe(!openConfirm);
    });
  });

  describe('handleConfirmDelete', () => {
    it('should call deletePostAndRefetch and openSnackbar', () => {
      setup();
      const confirmDialog = wrapper.find('[title="Delete"]');
      confirmDialog.simulate('confirm');
      expect(props.deletePostAndRefetch).toHaveBeenCalled();
      expect(props.openSnackbar).toHaveBeenCalled();
    });
  });
});
