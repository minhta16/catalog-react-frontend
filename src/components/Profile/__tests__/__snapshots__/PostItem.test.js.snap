// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/PostItem should render correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  justify="space-between"
  spacing={2}
>
  <ConfirmDialog
    contentText="Are you sure you want to delete bro?"
    onCancel={[Function]}
    onClose={[Function]}
    onConfirm={[Function]}
    open={false}
    title="Delete"
  />
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      direction="column"
    >
      <WithStyles(ForwardRef(Link))
        color="inherit"
        component={[Function]}
        exact="true"
        to="/1/undefined"
        underline="none"
        variant="h4"
      >
        bro
      </WithStyles(ForwardRef(Link))>
      <WithStyles(ForwardRef(Typography))
        variant="subtitle1"
      >
        Category: cat
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        variant="subtitle1"
      >
        Created: 2019
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        variant="caption"
      >
        Description: bro
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <WithStyles(ForwardRef(Typography)) />
    <WithStyles(ForwardRef(IconButton))
      onClick={[Function]}
    >
      <MoreVertIcon />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Menu))
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "bottom",
      }
    }
    getContentAnchorEl={null}
    id="simple-menu"
    keepMounted={true}
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <WithStyles(ForwardRef(MenuItem))
      component={[Function]}
      exact="true"
      onClick={[Function]}
      to="/profile/1/undefined/edit"
    >
      <WithStyles(ForwardRef(Grid))
        alignItems="center"
        container={true}
      >
        <UpdateIcon
          className="icon-padding"
          color="primary"
        />
        <WithStyles(ForwardRef(Typography))
          color="primary"
          variant="button"
        >
          Update
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      id="post-item-delete"
      onClick={[Function]}
    >
      <DeleteIcon
        className="icon-padding"
        color="error"
      />
      <WithStyles(ForwardRef(Typography))
        color="error"
        variant="button"
      >
        Delete
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Menu))>
</WithStyles(ForwardRef(Grid))>
`;
